lib/new-admin/graphql/types/users.type.js

```diff
@@ -0,0 +1,68 @@
+const typeDef = `
+  type UserSession {
+    sid: String!
+    sess: JSONObject!
+    expire: Date!
+  }
+
+  type User {
+    id: ID
+    username: String
+    role: String
+    enabled: Boolean
+    created: Date
+    last_accessed: Date
+    last_accessed_from: String
+    last_accessed_address: String
+  }
+
+  type TwoFactorSecret {
+    user_id: ID
+    secret: String!
+    otpauth: String!
+  }
+
+  type ResetToken {
+    token: String
+    user_id: ID
+    expire: Date
+  }
+
+  type RegistrationToken {
+    token: String
+    username: String
+    role: String
+    expire: Date
+  }
+
+  type Query {
+    users: [User] @auth(requires: [SUPERUSER])
+    sessions: [UserSession] @auth(requires: [SUPERUSER])
+    userSessions(username: String!): [UserSession] @auth(requires: [SUPERUSER])
+    userData: User
+    get2FASecret(username: String!, password: String!): TwoFactorSecret
+    confirm2FA(code: String!): Boolean @auth(requires: [SUPERUSER])
+    validateRegisterLink(token: String!): User
+    validateResetPasswordLink(token: String!): User
+    validateReset2FALink(token: String!): TwoFactorSecret
+  }
+
+  type Mutation {
+    deleteUser(id: ID!): User @auth(requires: [SUPERUSER])
+    deleteSession(sid: String!): UserSession @auth(requires: [SUPERUSER])
+    deleteUserSessions(username: String!): [UserSession] @auth(requires: [SUPERUSER])
+    changeUserRole(id: ID!, newRole: String!): User @auth(requires: [SUPERUSER])
```

Comment:
How hard would it be to move the 2fa to the same request as the actual change? Separate requests means that 2fa can be bypassed by just calling graphql directly

Reply:
Pinging to make sure you saw this one.

Reply:
It took some refactoring on both back and front-end, but I believe it is done at the first level we spoke yesterday